### PR TITLE
651 add eth methods validations

### PR DESCRIPTION
**Pull request by dimitrovmaksim**
**Description**:
Adds input validations for the `eth` methods. The validation are based on multiple rpc specifications and [openrpc.json](https://playground.open-rpc.org/?schemaUrl=https://raw.githubusercontent.com/hashgraph/hedera-json-rpc-relay/main/docs/openrpc.json&uiSchema%5BappBar%5D%5Bui:splitView%5D=false&uiSchema%5BappBar%5D%5Bui:input%5D=false&uiSchema%5BappBar%5D%5Bui:examplesDropdown%5D=false)

There are some cases where the validation is not as strict as Alchemy for example.

Also It adds temporary fixes to the acceptance tests, because they were inconsistent. There were various cases where hex values were sent without `0x` prefix, and even cases where hedera format (`0.0.xxx`) account, contract or transaction ID were sent as parameters.

I think it would be good to test this more extensively before releasing it as official image version as there may be some missing edge cases.

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-json-rpc-relay/issues/651

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: packages/server/tests/acceptance/erc20.spec.ts [hunk 1]
```diff
@@ -99,7 +99,7 @@ describe('@erc20 Acceptance Tests', async function () {
             });
 
             it('Relay can execute "eth_getCode" for ERC20 contract with evmAddress', async function () {
-                const res = await relay.call('eth_getCode', [contract.address], requestId);
+                const res = await relay.call('eth_getCode', [contract.address, "latest"], requestId);
```
Comment (Nana-EC) [inline packages/server/tests/acceptance/erc20.spec.ts:102]:
```suggestion
                const res = await relay.call('eth_getCode', [contract.address, 'latest'], requestId);
```

**Review by Nana-EC — COMMENTED** [review]
1st pass looking good. Will circle back for 2nd pass

File: packages/server/src/validator.ts [hunk 2]
```diff
@@ -0,0 +1,278 @@
+import { JsonRpcError, predefined } from '@hashgraph/json-rpc-relay';
```
Comment (georgi-l95) [inline packages/server/src/validator.ts:1]: Good progress. My opinion is that this file way too big, with way too many responsibilities. You can put all const variables in separated `constants.ts` file. Pull out all static methods in separate file as well and keep only one class per file. This way when 3rd person sees the files will know - okay, here are the constants, here we have this class, which does that and so on.. I personally try to aim for higher cohesion.

**Comment by dimitrovmaksim** [comment]
> 1st pass looking good. Will circle back for 2nd pass

Please also consider that I'm not fully sure about the Transaction object validations, there are various cases depending on the method or type of transaction and I can't find definitive answer how it should be, especially regarding required params

File: packages/server/src/server.ts [hunk 3]
```diff
@@ -21,6 +21,7 @@
 import { Relay, RelayImpl, JsonRpcError, predefined, MirrorNodeClientError } from '@hashgraph/json-rpc-relay';
```
Comment (georgi-l95) [inline packages/server/src/server.ts:21]: Nice, but my approach would be to make this validations in `koaJsonRpc` with the other checks. There we already have some kind of invalid params check, we can just extend it and not cluster `server.ts` with unnecessary code.
Overall looks good, it's important to have validations.

**Review by georgi-l95 — COMMENTED** [review]
LG, sharing my opinion.

File: packages/server/src/server.ts (hunk 3 above)
Comment (dimitrovmaksim) [reply packages/server/src/server.ts:21]: Yeah, I wasn't sure for the best place for this.

File: packages/server/src/validator.ts (hunk 2 above)
Comment (dimitrovmaksim) [reply packages/server/src/validator.ts:1]: The issue I see here is that this will probably cause circular dependencies and will probably require some level of code/flow refactoring to escape this.

File: packages/server/src/server.ts [hunk 4]
```diff
@@ -217,6 +218,18 @@ app.useRpc('eth_blockNumber', async () => {
  * returns: Gas used - hex encoded integer
  */
 app.useRpc('eth_estimateGas', async (params: any) => {
+  const validationError = Validator.validateParams(params, {
```
Comment (Nana-EC) [inline packages/server/src/server.ts:221]: Validating too early may prevent the log and metrics of the endpoint call from occurring.
Please check this is not the case.
if it is maybe you want to pass the Validator into the `logAndHandleResponse()` so that it gets run before the service logic is called

**Review by Nana-EC — COMMENTED** [review]
Nice validations. 
Please check they occur after initial entry point logs and metrics so we can still observe the calls but users get the response as to what they missed on the call

File: packages/server/src/server.ts (hunk 4 above)
Comment (dimitrovmaksim) [reply packages/server/src/server.ts:221]: If the validation happens in `logAndHandleResponse()`, we will only get part of the logs 
```
logger.debug(`${requestIdPrefix} ${methodName}`);
```
because trace logging the called method and the params passed
```
this.logger.trace(`${requestIdPrefix} getBlockByHash(hash=${hash}, showDetails=${showDetails})`);
```
happens in function of each method the `eth.ts` module in the `relay` package.
Would this be enough?

Comment (lukelee-sl) [reply packages/server/src/server.ts:221]: Good question.  Let me ask a question in response.  Does the validation error message give you enough information about all of the parameters were passed into the function and which one was failing validation?  If so, perhaps there is enough information for the reader of the log to piece together what went wrong?

One alternate suggestion which is kind of ugly as it involves a fair amount of refactoring is to pass in an object instead of the `requestId` string.  The object could contain the `validator` and the `requestId`.  I'm not sure what a good name for such an object might be.

Comment (dimitrovmaksim) [reply packages/server/src/server.ts:221]: It either gives you the index of the parameter, for example in
```
{ "method" : "eth_getBlockByNumber" ,"params" : ["1", false], "id": 1, "jsonrpc": "2.0" }
```
It will return that `param at index 0` is not a hex number.

In the case where one of the params itself is an object (Filter or Transaction)
```
{ "method" : "eth_getLogs", "params" : [ {"address": "0x1234"} ], "id":1 , "jsonrpc":"2.0" }
```
It will return that the `address` property of the `Filter` object is not a valid address

File: packages/server/tests/acceptance/rpc.spec.ts [hunk 5]
```diff
@@ -1067,7 +1067,7 @@ describe('@api RPC Server Acceptance Tests', function () {
             });
         });
 
-        describe('@release Unsupported RPC Endpoints', () => {
+        describe.only('@release Unsupported RPC Endpoints', () => {
```
Comment (Ivo-Yankov) [inline packages/server/tests/acceptance/rpc.spec.ts:1070]: Remove `.only`

**Review by Ivo-Yankov — APPROVED** [review]

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Review by natanasow — APPROVED** [review]
LG
